command-line/lib/Command/Address/Payment.hs

```diff
@@ -65,28 +59,48 @@ mod liftCmd = command "payment" $
             , indent 2 $ string ""
             , indent 2 $ bold $ string "$ cat addr.prv \\"
             , indent 4 $ bold $ string $ "| "<>progName<>" key public --with-chain-code \\"
-            , indent 4 $ bold $ string $ "| "<>progName<>" address payment --from-key --network-tag testnet"
+            , indent 4 $ bold $ string $ "| "<>progName<>" address payment --network-tag testnet"
             , indent 2 $ string "addr_test1vqrlltfahghjxl5sy5h5mvfrrlt6me5fqphhwjqvj5jd88cccqcek"
             ])
   where
     parser = Cmd
-        <$> credentialOpt
-        <*> networkTagOpt Shelley
+        <$> networkTagOpt Shelley
 
 run :: Cmd -> IO ()
-run Cmd{networkTag,credentialType} = do
-    case (mkNetworkDiscriminant . fromIntegral . unNetworkTag) networkTag of
-        Left ErrWrongNetworkTag{} -> do
-            fail "Invalid network tag. Must be between [0, 15]"
-        Right discriminant -> do
-            addr <- case credentialType of
-                CredentialFromKey -> do
-                    xpub <- hGetXPub stdin
-                    pure $ Shelley.paymentAddress discriminant (PaymentFromKey $ Shelley.liftXPub xpub)
-                CredentialFromScript -> do
-                    scriptHash <- hGetScriptHash stdin
-                    pure $ Shelley.paymentAddress discriminant (PaymentFromScript scriptHash)
-            B8.hPutStr stdout $ T.encodeUtf8 $ bech32With hrp addr
+run Cmd{networkTag} = do
+    discriminant <- fromNetworkTag networkTag
+    (hrp, bytes) <- hGetBech32 stdin allowedPrefixes
+    addr <- addressFromBytes discriminant bytes hrp
+    hPutBytes stdout (unAddress addr) (EBech32 addrHrp)
   where
-    hrp | networkTag == shelleyTestnet = [humanReadablePart|addr_test|]
-        | otherwise = [humanReadablePart|addr|]
+    addrHrp
+        | networkTag == shelleyTestnet = CIP5.addr_test
+        | otherwise = CIP5.addr
+
+    -- TODO: Also allow `XXX_vk` prefixes. We don't need the chain code to
+    -- construct a payment credential. This will however need some additional
+    -- abstraction over `xpubFromBytes` but I've done enough yake-shaving at
+    -- this stage, so leaving this as an item for later.
+    allowedPrefixes =
+        [ CIP5.addr_xvk
+        , CIP5.stake_xvk
+        , CIP5.script_xvk
```

Comment:
why do we allow `script_xvk` here? in credential from key either `addr_xvk` or `stake_xvk` can go, in credential from script we accept script hash which has `script` right?

Reply:
Hmmm. That is correct. script_xvk should not be there because they would be part of scripts.

Reply:
`script_xvkh` keys are building script which in the end when hashed should have `script`